source/internal-communication.rst

```diff
@@ -0,0 +1,539 @@
+     Kafka --> Component: OK
+   end
+
+The second phase starts after startup and its purpose is to process all
+messages received through Kafka. This will include any tenant that is created
+after all services are up and running. :numref:`Tenant bootstrapping` shows how
+to deal with these messages.
+
+.. _Tenant bootstrapping:
+.. uml::
+   :caption: Tenant bootstrapping
+   :align: center
+
+
+   control Kafka
+   control Component
+   control DataBroker
+
+   Kafka -> Component: MESSAGE\ntopic:98797ce98af...\nmessage: {"tenant" : "new-tenant"}
+   Component -> DataBroker: GET /topic/device-data\nHeaders: "Authorization: Bearer JWT"
+   note left
+     JWT contains
+     new tenant
+   end note
+   DataBroker --> Component: OK {"topic" : "876ca876g7..."}
+   Component -> Kafka: SUBSCRIBE\ntopic: 876ca876g7...
+   Kafka --> Component: OK
+   Component -> DataBroker: GET /topic/dojot.device-manager.devices\nHeaders: "Authorization: Bearer JWT"
+   note left
+     JWT contains
+     new tenant
+   end note
+   DataBroker --> Component: OK {"topic" : "22432c4a..."}
+   Component -> Kafka: SUBSCRIBE\ntopic: 22432c4a...
+   Kafka --> Component: OK
+
+
+All services that are somehow interested in using subjects should execute this
+procedure in order to correctly receive all messages.
+
+
+Auth + API gateway (Kong)
+-------------------------
+
+Auth is a service deeply connected to Kong. It is responsible for user
+management, authentication and authorization. As such, it is invoked by Kong
+whenever an request is received by one of its registered endpoints. This
+section will detail how this is performed and how they work together.
+
+Kong configuration
+++++++++++++++++++
+
+There are two configuration procedures when starting Kong within dojot:
+
+#. Migrating existing data
+#. Registering API endpoints and plugins.
+
+The first task is performed by simply invoking Kong with a special flag.
+
+The second one is performed by executing a configuration script
+`kong.config.sh`. Its only purpose is to register endpoints in Kong, such as:
+
+.. code-block:: bash
+
+
+    (curl -o /dev/null ${kong}/apis -sS -X POST \
+        --header "Content-Type: application/json" \
+        -d @- ) <<PAYLOAD
+    {
+        "name": "data-broker",
+        "uris": ["/device/(.*)/latest", "/subscription"],
+        "strip_uri": false,
+        "upstream_url": "http://data-broker:80"
+    }
+    PAYLOAD
+
+
+This command will register the endpoint `/device/*/latest` and `/subscription`
+and all requests to it are going to be forwarded to `http//data-broker:80`. You
+can check the documentation on how to add endpoints in `Kong's documentation`_.
+
+For some of its registered endpoints, `kong.config.sh` will add two plugins to
+selected endpoints:
+
+#. JWT generation. The documentation for this plugin is available at `Kong JWT
+   plugin page`_.
+#. Configuration a plugin which will forward all policies requests to Auth.
```

Comment:
We are missing an example of how to assign the endpoint to auth verification

Reply:
Added